pkg/workspace/workspace.go

```diff
@@ -156,74 +173,81 @@ func (ws *Workspace) CatalogPath(name *string) string {
-func (ws *Workspace) ListCatalogs() ([]*string, wfapi.Error) {
+func (ws *Workspace) ListCatalogs() ([]string, wfapi.Error) {
+	if !ws.isRootWorkspace {
+		return []string{""}, nil
+	}
 	catalogsPath := ws.CatalogBasePath()
 
 	_, err := fs.Stat(ws.fsys, catalogsPath)
 	if os.IsNotExist(err) {
 		// no catalogs directory, return an empty list
-		return []*string{}, nil
+		return []string{}, nil
 	} else if err != nil {
-		return []*string{}, wfapi.ErrorIo("failed to stat catalogs path", &catalogsPath, err)
+		return []string{}, wfapi.ErrorIo("failed to stat catalogs path", catalogsPath, err)
 	}
 
 	// list the directory
 	catalogs, err := fs.ReadDir(ws.fsys, catalogsPath)
 	if err != nil {
-		return []*string{}, wfapi.ErrorIo("failed to read catalogs dir", &catalogsPath, err)
+		return []string{}, wfapi.ErrorIo("failed to read catalogs dir", catalogsPath, err)
 	}
 
 	// build a list of subdirectories, each is a catalog
-	var list []*string
+	var list []string
 	for _, c := range catalogs {
-		if c.IsDir() {
+		if c.IsDir() && reCatalogName.MatchString(c.Name()) {
 			name := c.Name()
-			list = append(list, &name)
+			list = append(list, name)
 		}
 	}
 	return list, nil
 }
 
 // Get a catalog ware from a workspace, doing lookup by CatalogRef.
-// This will first check all catalogs within the "catalogs" subdirectory, if it exists
-// then, it will check the "catalog" subdirectory, if it exists
+// In a root workspace this will check valid catalogs within the "catalogs" subdirectory
+// In a non-root workspace, it will check the "catalog" subdirectory
 //
 // Errors:
 //
 //     - warpforge-error-io -- when reading of lineage or mirror files fails
 //     - warpforge-error-catalog-parse -- when ipld parsing of lineage or mirror files fails
 //     - warpforge-error-catalog-invalid -- when ipld parsing of lineage or mirror files fails
+//     - warpforge-error-internal -- when the catalog name is invalid, this shouldn't happen
 func (ws *Workspace) GetCatalogWare(ref wfapi.CatalogRef) (*wfapi.WareID, *wfapi.WarehouseAddr, wfapi.Error) {
 	// list the catalogs within the "catalogs" subdirectory
 	cats, err := ws.ListCatalogs()
 	if err != nil {
 		return nil, nil, err
 	}
 
-	// if it exists, add the "catalog" subdirectory to the end of the list
-	// this is done by adding a catalog with nil name, which refers to the unnamed catalog
-	// in the "catalog" subdirectory
-	catalogPath := filepath.Join(ws.rootPath, magicWorkspaceDirname, "catalog")
-	_, errRaw := fs.Stat(ws.fsys, catalogPath)
-	if errRaw == nil {
-		// "catalog" subdirectory exists, append nil
-		cats = append(cats, nil)
-	}
-
 	for _, c := range cats {
 		cat, err := ws.OpenCatalog(c)
 		if err != nil {
-			return nil, nil, err
+			switch err.(*wfapi.ErrorVal).Code() {
+			case "warpforge-error-catalog-name":
+				err := wfapi.ErrorInternal("referenced an invalid catalog", err)
+				return nil, nil, err
+			default:
+				// Error Codes -= warpforge-error-catalog-name
```

Comment:
@warpfork This might be a controversial decision. Not sure if you'd prefer it stay as a `warpforge-error-catalog-name` even though in this instance it really shouldn't happen because the names are already validated by `ListCatalogs`.

And I do it again [here](https://github.com/warpfork/warpforge/pull/27/files#diff-92c06c458c8332afdcac283317a34cb1302e1b722216cd26d10056c9c2bfb4c0R341-R349)

Reply:
This was updated to panic instead of returning an internal error. It still removes the error-catalog-name from the error output.

Reply:
SGTM.